### PR TITLE
feat: Remove support for 1.58 proc-macro abi

### PR DESCRIPTION
**Pull request by Veykril**
This seems old enough that we can drop the support for it now, the less ABIs we have the less work it is adjusting our span implementation.

Extracted from https://github.com/rust-lang/rust-analyzer/pull/14061, will rebase that over this once merged.

### CODE REVIEW

File: crates/proc-macro-srv/src/abis/mod.rs [hunk 1]
```diff
@@ -36,7 +35,6 @@ include!(concat!(env!("OUT_DIR"), "/rustc_version.rs"));
 pub(crate) use abi_sysroot::TokenStream as TestTokenStream;
 
 use super::dylib::LoadProcMacroDylibError;
-pub(crate) use abi_1_58::Abi as Abi_1_58;
 pub(crate) use abi_1_63::Abi as Abi_1_63;
```
Comment (lnicola) [inline crates/proc-macro-srv/src/abis/mod.rs:38]: You're next :grinning:!

**Comment by pravic** [comment]
While 1.58 is indeed old enough (more than [a year old](https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-1580-2022-01-13)), [the next supported one](https://github.com/rust-lang/rust-analyzer/tree/master/crates/proc-macro-srv/src/abis) is 1.63 which is only [6 months old](https://github.com/rust-lang/rust/blob/stable/RELEASES.md#version-1630-2022-08-11). And 1.62, for example, can't be used anymore with the latest RA.

**Comment by IniterWorker** [comment]
@lnicola How do you justify that rust-analyzer doesn't support old ABI, when RustC would?

**Comment by lnicola** [comment]
@IniterWorker what do you mean? rustc only supports proc macros built with the same version of it. You can't load proc macros compiled with 1.58 in rustc 1.67.

**Comment by IniterWorker** [comment]
Sorry for the miss direction with RustC! Related to https://github.com/rust-lang/rust-analyzer/issues/13046, I would like to understand why a `cargo check` will not generate any error with 1.59 definition in `rust-toolchain.toml` but rust-analyzer in VSCode will generate errors. Is it related? Or, I am wrong.

**Comment by lnicola** [comment]
With this PR, we dropped support for the proc macro ABI introduced in Rust 1.58, which was used until 1.63. If you're using Rust 1.59, you might want to disable proc macro support, or to stay on the previous RA release until you switch to a newer toolchain.

I can imagine upgrading is not feasible for you, but the RA support for old compilers was generally on a "best effort" basis.

**Comment by pravic** [comment]
> but the RA support for old compilers was generally on a "best effort" basis.

Well, removing the support of a 6 months old compiler is too fast, imho. But yeah, downgrading RA is an option.

**Comment by lnicola** [comment]
We did not remove support for a compiler, we removed support for the proc macro ABI it uses. You can still use RA to write code targeting old compilers, but without proc macro support. And if you're not using nightly, you can use a new compiler and test with an old one on CI.

This may change in the future, but it's pretty clearly [documented](https://rust-analyzer.github.io/manual.html#toolchain):

> Only **the latest** stable standard library source is officially supported for use with rust-analyzer. If you are using an older toolchain or have an override set, rust-analyzer may fail to understand the Rust source. You will either need to update your toolchain or use an older version of rust-analyzer that is compatible with your toolchain.

(emphasis mine)

The proc macro interface has always been a sore point and a pain to keep up. Fortunately, recent toolchains include our proc macro server, so this won't be a problem after we drop the 1.63 one too.

**Pull request merged**